### PR TITLE
[Final Project] From Allo to XLS: Bridging Hardware Accelerator DSLs Through Traditional Compilation to ASIC Backends

### LINKED ISSUE #607: Project Proposal: Allo -> ASIC Compiler Backend
# Project Proposal: Allo -> ASIC Compiler Backend

**What will you do?**

Build an ASIC-oriented backend for the Allo compiler that targets one (we haven't figured out which) of:

1. Siemens Catapult HLS: (extend the existing Allo -> Catapult path to support kernels currently unsupported), and/or
2. Google XLS: (create a new Allo -> XLS codegen path from scratch).

We’ll pick a few (like 1-3) tractable Allo kernels (e.g., 2D conv or other examples in https://github.com/cornell-zhang/allo/tree/main/examples) that are either (1) unsupported by the current Catapult backend or (2) clearly unsupported due to the absence of any XLS backend.

We want to build the following:
- A working backend capable of compiling the chosen kernels from Allo IR to the ASIC tool’s input (Catapult-friendly C++ with directives **or** XLS DSLX).
- Golden tests

**How will you do it?**
- Select 2–3 kernels that define a tractable subset of Allo to write a backend for
- Define each kernel’s Allo features used
  - For Catapult HLS: Generate Catapult-friendly C++/SystemC from Allo IR; emit pragmas (e.g., pipeline, unroll, memory/array partitioning) inferred from Allo’s scheduling metadata.
  - For Google XLS: Lower Allo IR to XLS IR/DSLX constructs
- Add a lowering pass for the target
- Write equivalence test benches that compare Allo reference outputs vs. target-sim outputs.

**How will you empirically measure success?**
- That our system builds/runs: The chosen nontrivial kernels compile and run end-to-end on the target.
- That our system is correct: Exact output equivalence against Allo golden references across randomized inputs and edge cases.
- That our system is somewhat fast/quality: To test quality, maybe we synthesize to RTL and measure area proxies and timing?

## Team members
- Cynthia Shao (@CynyuS)
- This is joint work with 2 other students in ECE 6775. However, we will clearly delineate which components were built by us.

### PR DESCRIPTION
**Pull request by CynyuS**
close #607

### CODE REVIEW

File: content/blog/2025-12-16-allo-xls-backend.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+Nikil Shyamsunder is an undergraduate junior studying Math and CS at Cornell. In his free time he likes trying to solve gerrymandering with algorithmic game theory, listen to Agatha Christie audio books, and go on scenic walks. 
+"""
+
+[[extra.authors]]
+name = "Cynthia Shao"
+[[extra.authors]]
+name = "Nikil Shyamsunder"
++++
+
+## What's the Goal?
```
Comment (sampsyo) [inline content/blog/2025-12-16-allo-xls-backend.md]: You don't really need to include the headings from my list of questions in the syllabus! Since this is for external consumption, you can use whatever structure works for your post. In particular, jumping right in with this intro without any kind of heading would work just fine here.

File: content/blog/2025-12-16-allo-xls-backend.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+Hardware accelerator design increasingly relies on portable compilation flows that can target diverse backends without extensive manual retargeting. This project explores building a compilation pathway from Allo, a Pythonic MLIR-based hardware accelerator DSL, to Google's XLS hardware synthesis ASIC backend.
```
Comment (sampsyo) [inline content/blog/2025-12-16-allo-xls-backend.md]: Please add links to Allo, MLIR, and XLS.

File: content/blog/2025-12-16-allo-xls-backend/_index.md
Comment (sampsyo) [inline content/blog/2025-12-16-allo-xls-backend/_index.md:1]: I could be wrong, but I think you want this to be `index.md` within the directory (rather than having an identical name to the directory) in order to be rendered correctly?

File: content/blog/2025-12-16-allo-xls-backend.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+Tiled, reordered, and pipelined variants that Allo can express. These introduced loop tiling and reordering transformations, as well as handling Allo kernel compositions. This demonstrated our backend could handle realistic optimization styles.
+
+**Stage 3: Small Fixed-Size Systolic Arrays**  
+A 2×2 grid of processing elements representing Allo's dataflow patterns. This required recognizing local data movement through FIFOs and mapping to XLS procs connected through channels.
+
+We originally wanted to commit to supporting: static loop nests, reductions, simple grid loops, temporary buffers, and basic arithmetic over various integer and float types. While we were able to hit most of these, our current implementation only supports `u32` for types in `fn` lowering, and due to limited support for memory integration XLS, we decided against handling memory optimizations. 
+
+## Implementation
+
+The full implementation is open source at [https://github.com/Nikil-Shyamsunder/allo-xls-backend/blob/main/allo/backend/xls/README.md].
```
Comment (sampsyo) [inline content/blog/2025-12-16-allo-xls-backend.md]: I think your Markdown here is missing the text for the link?

File: content/blog/2025-12-16-allo-xls-backend.md [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+### Systolic Array Generation (Method 1)
+
+XLS organizes computation as communicating processes based on Kahn process network semantics. A `proc` represents a stateful concurrent process with three components:
+- A `config` function declaring and wiring communication channels
+- An `init` function establishing initial state values  
+- A `next` function implementing a single state transition by consuming inputs, updating state, and producing outputs
+
+Rather than attempt general lowering from arbitrary Allo programs to explicit state machines (an intractably hard problem), we adopt a pattern-matching approach that specifically recognizes systolic array structures in MLIR and generates corresponding XLS grid implementations. We extracted metadata from the MLIR that drives a builder. The builder constructs DSLX proc ASTs for the PE and for the systolic grid that is then serialized.
+
+To see the generated XLS, see this link: https://github.com/Nikil-Shyamsunder/allo-xls-backend/tree/main/allo/backend/xls/examples/systolic
```
Comment (sampsyo) [inline content/blog/2025-12-16-allo-xls-backend.md]: Try using an actual Markdown link:

```
You can [see the generated XLS](https://github.com/Nikil-Shyamsunder/allo-xls-backend/tree/main/allo/backend/xls/examples/systolic) for one systolic array.
```

File: content/blog/2025-12-16-allo-xls-backend.md [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        ...
+    with allo.meta_elif(j == 0):
+        # Load A rows
+        ...
+    with allo.meta_else():
+        # Interior PE: MAC computation
+        ...
+```
+
+The resulting MLIR contains separate functions for each grid position: `gemm_0_0`, `gemm_0_1`, `gemm_1_0`, `gemm_1_1`, etc. Each function has different code depending on which `meta_if` branch was selected for that `(i, j)` coordinate at compile time. The challenge of this approach was that we had to reverse-engineer the spatial structure and PE types from this unrolled representation by analyzing the behavior of each generated function.
```
Comment (sampsyo) [inline content/blog/2025-12-16-allo-xls-backend.md]: Hmm, I don't quite understand why you need to "reverse-engineer the spatial structure and PE types." First of all, I'm not exactly sure what "spatial structure" means here… but assuming it's something about the 2D layout of the PEs, I'm also not sure why that matters. I guess I'd assume that the PEs already communicate through some manner of communication channel, and that's all you need to "link them up" according to their data exchange?

File: content/blog/2025-12-16-allo-xls-backend.md [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+To solve this, we utilized pattern matching using `SystolicDetector.is_metaif_systolic()` method by searching for multiple `gemm_i_j` functions, functions containing multiply-accumulate logic (interior PEs), and functions with only loads or stream operations (edge handlers). A meta-if systolic array is detected when there are ≥4 `gemm_i_j` functions and at least one contains MAC computation. The `MetaIfSystolicTranslator` analyzes each unrolled function to classify its role (MAC PEs, input loaders, and drain PEs). From the function names, we extract grid dimensions. We also extract the K loop bound from interior PE functions. Once we have this metadata, we plug this into the builder from Method 1.
```
Comment (sampsyo) [inline content/blog/2025-12-16-allo-xls-backend.md]: Ah, I see: it's because you want to do the pattern-recognition thing again, rather than directly translating the Allo PEs into XLS PEs. It would be good to clarify this up front.

File: content/blog/2025-12-16-allo-xls-backend/_index.md [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
...tically debug by isolating which XLS passes were failing, then examining the intermediate IR representations and optimizations applied.  This debugging process was time-consuming and required deep understanding of XLS's internal pipeline stages. Documentation was sparse for these advanced features, so we relied on reading XLS source code and experimenting with minimal test cases. Professor Sampson's pointer to the [XLS matmul_4x4 example](https://github.com/google/xls/blob/main/xls/examples/matmul_4x4/matmul_4x4.x) and the [proc tutorial](https://google.github.io/xls/tutorials/how_to_use_procs/#channel-arrays-and-loop-based-spawning) proved invaluable for understanding channel arrays and loop-based spawning patterns. In the midst of lowering the designs to Verilog, we found a bug in the default fifo configurations, which was fixed the next day! But, during the process of this project we found out that the golden matmul example couldn't properly lower to Verilog, as XLS's scheduling created circular dependencies within their fifo configuration. Thus, we had to create our own XLS systolic array example by playing around with DSLX and seeing what could actually get lowered to Verilog.
```
Comment (sampsyo) [inline content/blog/2025-12-16-allo-xls-backend/_index.md:101]: It would be really really good to report this bug in the DSLX systolic array example to the XLS team! I'm sure they would appreciate it.

File: content/blog/2025-12-16-allo-xls-backend/_index.md [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+### Verilator Simulation Infrastructure
+
+Setting up cycle-accurate simulation required building a complete testbench infrastructure. We created generic design under test (DUT) files for Verilator, and scripts to run the Verilator test file under different matrix sizes and codegen with different pipeline stages. We additionally implemented cycle-by-cycle output checking, and added scripts to generate CSVs to output the cycles in a digestible manner. For systolic arrays, this was particularly complex because we needed to feed matrix data into edge channels following the correct protocol, respect valid/ready handshaking on FIFO channels, collect results at the right time after K iterations complete, and verify correctness despite cycle-level timing variations. Getting the timing right required careful analysis of the generated Verilog to understand XLS's channel implementation and synchronization behavior.
```
Comment (sampsyo) [inline content/blog/2025-12-16-allo-xls-backend/_index.md:105]: Yep, this does indeed sound like a lot of work…

File: content/blog/2025-12-16-allo-xls-backend.md [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+**Scalability Implications:**
+
+For 2×2 matrices, combinational designs dominate, at least with our XLS systolic array design. However, extrapolating to larger matrices there may be an inflection point where systolic arrays begin to dominate.
+
+## Future Directions
+1. Handling more Allo Constructs: in particular, more of the Allo customization and scheduling primitives, particularly those having to do with memory.
+
+2. Extending Pattern Coverage: the current systolic array pattern matcher could be extended to recognize other regular structures, or generalized to handle more complex systolic layouts and massage them into `proc`s. 
+
+3. Choosing between `fn`s and `proc`s : Currently, our `fn` passes and our `proc` passes our disjoint. It would be nice if a user could specify what they want to lower into pure dataflow `fn`s vs. what they want to be a proc. Perhaps we could allow a pre-codegen flag in the MLIR that a user can specify to make this decision. Or we can use loop trip count analysis to decide between unrolling and proc-based iteration.
```
Comment (sampsyo) [inline content/blog/2025-12-16-allo-xls-backend.md]: Or, I suppose, you could just try to `fn` everything you can `fn` and only fall back to `proc` when that's not possible: e.g., for loops with dynamic trip counts.

File: content/blog/2025-12-16-allo-xls-backend/_index.md [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+```rust
+#[quickcheck(test_count=5)]
+fn prop_gemm_equivalence(A: s32[32][32], B: s32[32][32]) -> bool {
+    let result_compiler = allo_gemm_compiler(A, B);
+    let result_reference = allo_gemm_reference(A, B);
+    result_compiler == result_reference
+}
+```
+
+The XLS interpreter automatically generates randomized 32×32 integer matrices as inputs, executes both implementations, and verifies bitwise equality of outputs. This provides substantially greater input space coverage than hand-crafted test cases. Our framework successfully validates the `simple`, `split_outer`, and `split_both_merge` benchmarks, with all random trials passing. The `split_inner` and `split_both` benchmarks currently fail differential testing, correctly finding bugs, we realized, in our LLM’s implementations!
```
Comment (sampsyo) [inline content/blog/2025-12-16-allo-xls-backend/_index.md:210]: That's pretty wild! Very cool that the XLS QuickCheck stuff was so useful in this scenario.

**Review by sampsyo — CHANGES_REQUESTED** [review]
Very cool, y'all—it's awesome that you got this working in an end-to-end way, even if it does use a brittle "pattern recognition" approach for now. And it's cool to see the generated ASIC layouts. :)

I have a few low-level writing suggestions—could you please address those and then use GitHub's "request review" button to let me know when you're ready to publish?

**Comment by sampsyo** [comment]
Just a friendly reminder to please address a few of my writing comments above so we can publish your post.

File: content/blog/2025-12-16-allo-xls-backend/_index.md (hunk 7 above)
Comment (Nikil-Shyamsunder) [reply content/blog/2025-12-16-allo-xls-backend/_index.md:101]: Will do!

**Comment by Nikil-Shyamsunder** [comment]
Hi professor, I just implemented the requested changes. Sorry about the wait. Happy holidays! I don't see a way for me to "request review" - I think Cynthia might be able to do it because she is the author of the PR, but she's in a timezone outside of working hours right now. Hopefully she can get to it tomorrow morning. Thanks

**Comment by sampsyo** [comment]
No worries! This looks great. Have a lovely break!
